src/agent/agentService.ts

```diff
@@ -98,80 +133,88 @@ function createAgentService({
+      const body = Buffer.concat(clientBodyBuffers);
+      const meta = call.metadata;
+      const vaultNameOrId = meta.get('vaultNameOrId').pop()!.toString();
+      if (vaultNameOrId == null)
+        throw new ErrorGRPC('vault-name not in metadata');
+      let vaultId = await vaultManager.getVaultId(vaultNameOrId as VaultName);
+      if (!vaultId) {
         try {
-          vaultId = makeVaultId(vaultNameOrId);
-          await vaultManager.openVault(vaultId);
+          vaultId = vaultsUtils.makeVaultId(vaultNameOrId);
         } catch (err) {
-          if (
-            err instanceof vaultsErrors.ErrorVaultUndefined ||
-            err instanceof SyntaxError
-          ) {
-            vaultId = await vaultManager.getVaultId(vaultNameOrId as VaultName);
-            await vaultManager.openVault(vaultId);
-          } else {
-            throw err;
-          }
+          await genDuplex.throw(new vaultsErrors.ErrorVaultsVaultUndefined());
+          return;
         }
-        // TODO: Check the permissions here
-        const response = new vaultsPB.PackChunk();
-        const [sideBand, progressStream] = await vaultManager.handlePackRequest(
-          vaultId,
-          Buffer.from(body),
-        );
-        response.setChunk(Buffer.from('0008NAK\n'));
-        await write(response);
-        const responseBuffers: Buffer[] = [];
-        await new Promise<void>((resolve, reject) => {
-          sideBand.on('data', async (data: Buffer) => {
-            responseBuffers.push(data);
-          });
-          sideBand.on('end', async () => {
-            response.setChunk(Buffer.concat(responseBuffers));
-            await write(response);
-            resolve();
-          });
-          sideBand.on('error', (err) => {
-            reject(err);
-          });
-          progressStream.write(Buffer.from('0014progress is at 50%\n'));
-          progressStream.end();
+      }
+      const response = new vaultsPB.PackChunk();
+      const [sideBand, progressStream] = await vaultManager.handlePackRequest(
+        vaultId,
+        Buffer.from(body),
+      );
+      response.setChunk(Buffer.from('0008NAK\n'));
+      await genDuplex.write(response);
+      const responseBuffers: Uint8Array[] = [];
+      await new Promise<void>((resolve, reject) => {
+        sideBand.on('data', async (data: Uint8Array) => {
+          responseBuffers.push(data);
+        });
+        sideBand.on('end', async () => {
+          response.setChunk(Buffer.concat(responseBuffers));
+          await genDuplex.write(response);
+          resolve();
+        });
+        sideBand.on('error', (err) => {
+          reject(err);
         });
-        call.end();
+        progressStream.write(Buffer.from('0014progress is at 50%\n'));
+        progressStream.end();
       });
+      await genDuplex.next(null);
     },
-    vaultsScan: async (
-      call: grpc.ServerWritableStream<nodesPB.Node, vaultsPB.Vault>,
+    nodesScan: async (
```

Comment:
Should be `vaultsScan` with the ability to take gestalt id.

Reply:
As far as I know this just scans a single node and not the whole gestalt. gestaltId applicable here?

Reply:
It can be a gestalt ID in the future. For now stick with a node Id. Because you can scan the entire target gestalt by broadcasting it and following the crytolinks in the gestalt graph.